aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,877 @@
+    create_general_description_(description_);
+  }
+
+  ~csrOutput() {
+    TORCH_CUDASPARSE_CHECK(cusparseDestroyMatDescr(description_));
+  }
+
+  int size(int index) const {
+    return size_.at(index);
+  }
+};
+
+template<class scalar_t> 
+struct csrMatrixRef {
+  int* csr_indices_{nullptr};
+  int* csr_pointers_{nullptr};
+  scalar_t* csr_values_{nullptr};
+  int nnz_{0};
+  std::vector<int> size_{};
+
+  cusparseMatDescr_t description_{0};
+
+  csrMatrixRef() {
+    create_general_description_(description_);
+  }
+
+  csrMatrixRef(
+      int* csr_indices,
+      int* csr_pointers,
+      scalar_t* csr_values,
+      int nnz,
+      const std::vector<int>& size)
+      : csr_indices_{csr_indices},
+        csr_pointers_{csr_pointers},
+        csr_values_{csr_values},
+        size_{size} {
+    nnz_ = nnz;
+    create_general_description_(description_);
+  }
+
+  ~csrMatrixRef() {
+    TORCH_CUDASPARSE_CHECK(cusparseDestroyMatDescr(description_));
+  }
+ 
+  int size(int index) const {
+    return size_.at(index);
+  } 
+};
+
+using DcsrMatrixRef = csrMatrixRef<double>;
+using ScsrMatrixRef = csrMatrixRef<float>; 
+
+#if IS_CUSPARSE11_AVAILABLE()
+
+cudaDataType getTensorCudaDataType(Tensor self) {
+  cudaDataType cuda_data_type;
+  switch (self.scalar_type()) {
+    case ScalarType::Float:
+      cuda_data_type = CUDA_R_32F;
+      break;
+    case ScalarType::Double:
+      cuda_data_type = CUDA_R_64F;
+      break;
+    default:
+      TORCH_CHECK(false, "Tensor types must be either float32 or float64");
+      break;
+  }
+  return cuda_data_type;
+}
+
+template <class scalar_t>
+csrOutput cuSparse_matrix_multiply(
+    const csrMatrixRef<scalar_t>& A,
+    const csrMatrixRef<scalar_t>& B,
+    Tensor& output_values,
+    IntTensor& output_indices) {
+  
+  cudaDataType cuda_data_type = getTensorCudaDataType(output_values);
+
+  const int A_num_rows = A.size(0);
+  const int A_num_cols = A.size(1);
+  const int A_num_nnz = A.nnz_;
+
+  const int B_num_rows = B.size(0);
+  const int B_num_cols = B.size(1);
+  const int B_num_nnz = B.nnz_;
+
+  int* dA_csrOffsets = A.csr_pointers_;
+  int* dA_columns = A.csr_indices_;
+  scalar_t* dA_values = A.csr_values_;
+
+  int* dB_csrOffsets = B.csr_pointers_;
+  int* dB_columns = B.csr_indices_;
+  scalar_t* dB_values = B.csr_values_;
+
+  csrOutput out({A.size(0), B.size(1)});
+
+  out.csr_pointers_ = at::empty({out.size(0) + 1}, output_indices.options().dtype(kInt));
+
+  int* dC_csrOffsets = out.csr_pointers_.data_ptr<int>();
+  int* dC_columns = nullptr;
+  scalar_t* dC_values = nullptr; 
+
+  scalar_t alpha = 1.0f;
+  scalar_t beta = 0.0f;
+  cusparseOperation_t opA = CUSPARSE_OPERATION_NON_TRANSPOSE;
+  cusparseOperation_t opB = CUSPARSE_OPERATION_NON_TRANSPOSE;
+  cudaDataType computeType = cuda_data_type;
+
+  //--------------------------------------------------------------------------
+  // CUSPARSE APIs
+  cusparseHandle_t handle = NULL;
```

Comment:
please get cusparse handles from the pool, getCurrentCUDASparseHandle. This will take care of not leaking the handles, and also of setting the correct stream on the handle